### PR TITLE
job-runner: testing runs isolated to their own container

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Introduce job-runner, a new asyncio python application designed to replace a good chunk of our backend CI infra:
 - `tests-invoke`
 - `make-checkout`
 - `s3-streamer`
 - (eventually, maybe) `run-queue`

and dramatically simplify others:
 - `issue-scan`
 - `tests-scan`

while making it easier to play with our testing infrastructure locally.

The main new feature vs. the existing stuff is that each testing job is run in its own one-time-use container.  The image for that container is under control of the revision being tested, allowing for gating of task container upgrades (avoiding the usual wave of broken pixels, linting changes, etc.).

Instead of running our jobs from the giant blobs of shell pasted together in `issue-scan` or `tests-scan` we consume them as the new "Job" json format, already introduced in #5932.

The new approach also makes a conscious effort to reduce the amount of exposure of secrets to test runs (although we have some more work to do here).

TODO:
 - [x] Includes #6018 to fix tests
 - [x] Needs https://github.com/cockpit-project/cockpituous/pull/586 ; once that lands, revert the "TEMP HACK" commit
 - [x] Needs https://github.com/cockpit-project/cockpituous/pull/583 **deployed to prod** (it isn't yet) before  landing, so that it will actually work
 - [x] Needs https://github.com/cockpit-project/cockpituous/pull/588 to ensure we don't break image refreshes
 - [x] [Add host name to log](https://github.com/cockpit-project/bots/pull/6017#issuecomment-1980207904), like the old s3-streamer
 - [ ] ~cidfile crash after successful runs https://paste.centos.org/view/01fcdfc8 with `podman-remote`; this breaks attachments, thus a blocker~ not reproducible, ignore

### CODE REVIEW

File: job-runner [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                sys.exit(f'Poorly formed job: {exc}')
+
+        case other:
+            assert_never(other)
+
+    async with JobContext(args.config_file) as ctx:
+        await run_job(job, ctx)
+
+if __name__ == '__main__':
+    asyncio.run(main())
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply job-runner:86]: lolwat

**Comment by allisonkarlitskaya** [comment]
Random note: while hacking on this, despite all kinds of exceptions and irregular exits for all kinds of reasons (including many programming errors), I haven't seen a single leaked container in over two weeks (when I got the `finally:` block working properly to remove the container image).

**Comment by martinpitt** [comment]
@allisonkarlitskaya If you want to get this green, how about adding `branch: runner-toml` to https://github.com/cockpit-project/bots/blob/job-runner/.github/workflows/test.yml#L67 temporarily? In a separate "TEST:" commit of course. Then we can co-develop them.

**Comment by martinpitt** [comment]
I did that "test against cockpitous branch" temp hack now, and also noticed that I broke bots tests (see PR #6018). I included that fix here.

The test gets far now, but the "inner" bots unit test [fails on](https://github.com/cockpit-project/bots/actions/runs/8161308519/job/22309781120?pr=6017#step:8:330) "Failure: cockpit-project/bots#1 is closed", which sounds like a problem with my mock github API. That's a problem for tomorrow.

**Comment by martinpitt** [comment]
Rebased after landing #6018

File: job-runner.toml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+[container]
+command = 'podman'
+args = []
+run-args = ['--device=/dev/kvm']
+default-image = 'quay.io/cockpit/tasks:latest'
+
+[container.secrets]
+# see podman-secret(1)
+# github-token = '--secret=github-token'
```
Comment (martinpitt) [inline job-runner.toml:11]: As every human dev needs ~/.config/cockpit-dev/github-token token anyway, let's add an example how to use a normal file-based one? TBH I don't really see the point of creating a podman secret for this. It's just duplication and extra work to set it up.

I have
```
github-token='--volume=/home/martin/.config/cockpit-dev/github-token:/run/secrets/webhook/.config--github-token:U,ro'
```

and this is actually quite awkward: There is little choice about the host path (and moreover, this doesn't accept `~` or `$XDG_CONFIG_DIR`), and our infra already defines where the container expects the secret.

I know this has to be set on prod infra (`/var/lib/cockpit-tasks/secrets/...`), but that can happen in cockpituous `install-service`.

File: job-runner.toml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+[logs]
+driver='local'   # 's3' or 'local'
+
+[logs.s3]
+# hint: podman run --rm --net=host quay.io/minio/minio server /var
+url = 'http://127.0.0.1:9000/tmp/'
+key = {access='minioadmin', secret='minioadmin'}
+user-agent = 'job-runner (cockpit-project/bots)'
```
Comment (martinpitt) [inline job-runner.toml:20]: This seems uninteresting, can this become/stay an internal default? And commented out here if you care?

File: job-runner.toml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+link = 'http://127.0.0.1:8000/'
+
+[forge]
+driver='github'
+
+[forge.github]
+clone-url = 'https://github.com/'
+api-url = 'https://api.github.com/'
+post = true  # whether to post statuses, open issues, etc.
+user-agent = 'job-runner (cockpit-project/bots)'
```
Comment (martinpitt) [inline job-runner.toml:35]: Dito

File: job-runner.toml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+# (at least) one of `token` or `post = false` must be set
+# token = 'ghp_XXX'
```
Comment (martinpitt) [inline job-runner.toml:37]: That feels redundant. job-runner itself should just default to `${COCKPIT_GITHUB_TOKEN_FILE:-~/.config/cockpit-dev/github-token}`, in other words, use the `task.github` API. And the spawned temp containers would use `container.secrets.github-token`. Why is this needed here *again*?

File: lib/aio/drivers.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+# Copyright (C) 2024 Red Hat, Inc.
```
Comment (martinpitt) [inline lib/aio/drivers.py:1]: This is only used in one place -- could it be merged into abc.py or jobcontext.py? That'd avoid a lot of boilerplate.

File: lib/aio/github.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                return Subject(clone_url, get_str(get_dict(pull, 'head'), 'sha'), target)
+
+        else:
+            sha = await resolve_branch(self.session, clone_url, branch)
+            return Subject(clone_url, sha, target)
+
+
+class GitHubStatus(Status):
+    def __init__(self, api: GitHub, repo: str, revision: str, context: str, link: URL) -> None:
+        logger.debug('GitHub repo %s context %s link %s', repo, context, link)
```
Comment (martinpitt) [inline lib/aio/github.py:166]: why not `revision`? That's no less important than repo and context.

File: lib/aio/job.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
+# Copyright (C) 2024 Red Hat, Inc.
```
Comment (martinpitt) [inline lib/aio/job.py:1]: What's the rationale of splitting this out? This is core functionality of `job-runner`, and not being imported anywhere else. There also isn't a unit test for it, nor does it seem realistic to have one.

**Review by martinpitt — CHANGES_REQUESTED** [review]
First round of review. We need to talk :grin:

File: lib/aio/job.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        async with spawn(args, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.STDOUT) as container:
+            try:
+                # Log data until we hit EOF
+                assert container.stdout is not None
+                async for block in read_utf8(container.stdout):
+                    log.write(block)
+
+                # Upload all attachments
+                # TODO: live updates
+                # TODO: invent async tarfile for StreamReader
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/job.py:116]: I.e. I'd much rather just start 4 parallel `run-queue`s and have everything be sync. This retains linear (not interleaved) logs, is more robust when crashes happen (not kill all parallel jobs), and we can more easily scale the containers up and down. Running four monitor containers in parallel is a negligible runtime resource cost.

**Comment by martinpitt** [comment]
[test_pr() failed](https://github.com/cockpit-project/bots/actions/runs/8167407070/job/22327723735?pr=6017#step:8:669
). The inner bots tests succeeded (and the statuses bubble is green here), but the output check failed:
```
ERROR: did not find 'Running on:.*cockpituous' in '[test log]`
```

That used to be done by the old s3-streamer:
```
s3-streamer:                Running on:    {platform.node()}
```

While we can of course change the string format, the node where `job-runner` runs on is vitally important for maintenance. So this needs to be put back. I'll add a TODO entry to the description for this.

**Comment by martinpitt** [comment]
I adjusted the expected test output in https://github.com/cockpit-project/cockpituous/pull/587 and added a FIXUP commit here which adds back 'Running on:". The integration test passes now.

File: lib/aio/job.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+            f'--cidfile={cidfile}',
+            '--interactive',
+            *(f'--env={k}={v}' for k, v in job.env.items()),
+            '--env=TEST_ATTACHMENTS=/tmp/attachments',
+            f'--env=COCKPIT_CI_LOG_URL={log.url}',
+            *(arg for name, arg in ctx.secrets_args.items() if name in job.secrets),
+            job.container or ctx.default_image,
+            'python3', '-c',
+            (BOTS_DIR / 'checkout-and-run').read_text(),  # lulz
+            f'--revision={subject.sha}',
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py:96]: I was looking to avoid the usual selinux issues, and without having to relabel.

I'd be okay copying the script to the already existing temporary directory and letting the container relabel the copy?

File: checkout-and-run [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    sys.exit('\n*** Failed to execute entry point.')
+
+
+def main() -> None:
+    # All output to stdout — otherwise podman reorders things
+    os.dup2(1, 2)
+
+    log('\n+ [checkout-and-run]', shlex.join(sys.argv[1:]))
+
+    parser = argparse.ArgumentParser()
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply checkout-and-run:105]: I'm having trouble thinking of anything.  Do you have a suggestion?

File: checkout-and-run [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+def git(*cmd: str, tries: int = 1) -> None:
+    run('git', *cmd, tries=tries)
+
+
+def git_output(*cmd: str) -> str:
+    return output('git', *cmd)
+
+
+def checkout_and_run(repository: str, revision: str | None, rebase: str | None, command: Sequence[str]) -> None:
+    run('cat', '/usr/lib/os-release')
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply checkout-and-run:48]: Let's not have this argument.  I'll erase it. :)

File: checkout-and-run [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    run('cat', '/run/.containerenv')
+    run('hostname')
+    run('uname', '-a')
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply checkout-and-run:51]: I figured the kernel version would be useful info because that's the one thing not under control of the container image...

File: lib/aio/job.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/job.py:96]: Ah, SELinux -- if it's easy and nice with a temp dir copy, do that, otherwise please just close this conversation. Not important.

File: lib/aio/gitresolve.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
+# Copyright (C) 2024 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/gitresolve.py:1]: Gone.  Using the REST API for this now.  We have to do two calls, but it's an awful lot less code.

File: lib/aio/job.py (hunk 8 above)
Comment (allisonkarlitskaya) [reply lib/aio/job.py:1]: I want to unit test this stuff and I also want to call it from a future amqp-fed mode (which will likely live in a separate file).

File: lib/aio/job.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py:117]: I think we resolved this in the meeting?

File: checkout-and-run (hunk 11 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply checkout-and-run:105]: :thinking:  "Check out a git repo and run a command, with minimal environment assumptions; mainly used by job-runner"

?

File: lib/aio/local.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
+# Copyright (C) 2024 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/local.py:1]: Addressed in the meeting.

For the record I think an actually useful log backend to add would be one that writes directly to stdout.  That's for the future, though.

File: checkout-and-run (hunk 13 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply checkout-and-run:51]: OK, let's keep that then, thanks!

File: lib/aio/jsonutil.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
+# Copyright (C) 2023 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/jsonutil.py:1]: This is less about untrusted input and more about convincing mypy that we're not making mistakes.  That's how this stuff came about in the first place in the cockpit bridge as well (which also doesn't deal with untrusted inputs).

File: lib/aio/github.py [hunk 17]
```diff
@@ -0,0 +1,179 @@
+# Copyright (C) 2024 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/github.py:1]: Discussed online -- I'm willing to swallow that bit.

File: lib/aio/job.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/job.py:117]: Discussed this morning, this is a rather binary discussion and I suppose you won that part under the rules of meritocracy. 

I'll still miss the live attachments, though :cry:

File: lib/aio/jsonutil.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/jsonutil.py:1]: Yeah, I get it in the bridge, but we don't need to be concerned about untrusted input here. I like mypy, but having to have such overhead for accessing a simple toml file feels too expensive to me.

File: checkout-and-run (hunk 11 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply checkout-and-run:105]: that's long :)

[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply checkout-and-run:105]:
```
usage: checkout-and-run [-h] [--revision REVISION] [--rebase REBASE] repository [command ...]

Check out a git repo and run a command; mainly used by job-runner
```

File: checkout-and-run (hunk 13 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply checkout-and-run:51]: Erased all the others.

File: checkout-and-run [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+def run(*cmd: str, tries: int = 1) -> None:
+    joined = shlex.join(cmd)
+
+    for attempt in range(1, tries + 1):
+        try:
+            log('\n+', joined)
+            subprocess.check_call(cmd)
+            return
+        except subprocess.CalledProcessError as exc:
+            log(f'\n> Attempt {attempt} failed with code {exc.returncode}.')
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply checkout-and-run:27]: I added a simple `2 ** attempt` here.  It also waits after the last failure, but who cares — at that point we're gonna die anyway.

File: job-runner.toml [hunk 19]
```diff
@@ -0,0 +1,37 @@
+# Install to ~/.config/cockpit-dev/job-runner.toml
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply job-runner.toml:1]: Expanded this considerably.

File: job-runner.toml [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+[container]
+command = 'podman'
+args = []
+run-args = ['--device=/dev/kvm']
+default-image = 'quay.io/cockpit/tasks:latest'
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply job-runner.toml:7]: Discussed in the meeting.  This file *is* the internal defaults.  Hopefully the new comment in the header clears that up.

File: job-runner.toml (hunk 2 above)
Comment (allisonkarlitskaya) [reply job-runner.toml:11]: Two reasons against this:
 - the human user may well want to use job-runner with a different token than they use for themselves.  For me that's my personal GitHub account vs. `cockpit-copilot` which I use as my testing account, for example.  One of those has admin access to all things cockpit, and the other (as a testing account) very much does it.
 - the usual issues with selinux labels

File: job-runner.toml [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+[container.secrets]
+# see podman-secret(1)
+# github-token = '--secret=github-token'
+
+[logs]
+driver='local'   # 's3' or 'local'
+
+[logs.s3]
+# hint: podman run --rm --net=host quay.io/minio/minio server /var
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply job-runner.toml:17]: Ya.  This is done for exceptionally evil reasons: `/var` was chosen specifically because it already contains a world-writable directory called `tmp`, which `minio` is then prepared to serve as a bucket called `tmp`.

This is the only way I could find to get a semi-functional config out of the box without needing to use `mc`.  It's also not completely functional, though: you can't read anything without a token.  In general I strongly prefer either a real-world S3 bucket ([play.min.io](https://play.min.io:9443/) is interesting if you don't otherwise have something you can use) or the default `local` backend along with `python -m http.server`.

File: job-runner.toml (hunk 3 above)
Comment (allisonkarlitskaya) [reply job-runner.toml:20]: See above notes about internal defaults.

File: job-runner.toml (hunk 5 above)
Comment (allisonkarlitskaya) [reply job-runner.toml:37]: This is worth discussing, at least as a fallback.  Ditto for S3 keys.  Let's keep that in mind for a follow-up.

File: lib/aio/drivers.py (hunk 6 above)
Comment (allisonkarlitskaya) [reply lib/aio/drivers.py:1]: It can't be merged into `abc.py` (due to the cyclical dependencies that would result) but it can definitely be merged into `jobcontext.py` if you feel strongly about it.

File: lib/aio/github.py (hunk 7 above)
Comment (allisonkarlitskaya) [reply lib/aio/github.py:166]: OK.

File: lib/aio/jsonutil.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/jsonutil.py:1]: It's also for the job objects which we read off of the network.

Also keep in mind, the way that we define the job object, for example, these helpers actually help us get the types right in the first place.  Consider:

```
class Job:
    def __init__(self, obj: JsonObject) -> None:
        # test subject specification
        self.repo = get_str(obj, 'repo')
        self.sha = get_str(obj, 'sha', None)
        self.pull = get_int(obj, 'pull', None)
        self.branch = get_str(obj, 'branch', None)
        self.target = get_str(obj, 'target', None)

 ...etc...
```

mypy only knows the types of those attributes because of `get_str()` et al. being correctly typed.  This has found real issues during refactoring and in general has been a big help...

File: lib/aio/job.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py:96]: I did the copy.  It's working OK.

File: lib/aio/jsonutil.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/jsonutil.py:1]: OK, resolving..

File: job-runner.toml (hunk 21 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply job-runner.toml:17]: Hah, thansk for the explanation :nerd_face:

File: job-runner.toml (hunk 2 above)
Comment (martinpitt) [reply job-runner.toml:11]: Hmkay (you have a high privileged .config/cockpit-dev/github-token ? That sounds highly dangerous..)

Could you add a full example then?

File: checkout-and-run [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+    run('uname', '-a')
+    run('mkdir', '-p', os.environ['TEST_ATTACHMENTS'])
+
+    git('clone', '--', repository, 'make-checkout-workdir', tries=5)
+
+    print('\n+ cd make-checkout-workdir', file=sys.stderr)
+    try:
+        os.chdir('make-checkout-workdir')
+    except OSError as exc:
+        sys.exit(str(exc))
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply checkout-and-run:60]: Sure.

File: lib/aio/drivers.py (hunk 6 above)
Comment (allisonkarlitskaya) [reply lib/aio/drivers.py:1]: Done.

File: .github/workflows/test.yml [hunk 23]
```diff
@@ -19,6 +19,39 @@ jobs:
       - name: Run test
         run: test/run
 
+  job-runner:
```
Comment (martinpitt) [inline .github/workflows/test.yml:22]: Is this meant to be temporary, or do you want to keep this? Run-local's `test_pr()` covers the same job-runner and bots PR test, except it also covers the machinery around it.

(No strong opinion, just curious)

**Comment by martinpitt** [comment]
[This failure looks very relevant](https://github.com/cockpit-project/bots/actions/runs/8176057372/job/22354604776?pr=6017#step:8:304):
```
Error: statfs /tmp/tmp7wu7vgej/checkout-and-run: no such file or directory
```
Fortunately it already happens in test_mock_pr(), and it effortlessly reproduces in `COCKPIT_BOTS_BRANCH=job-runner tasks/run-local.sh`. Will have a look tomorrow morning unless you beat me to it.

File: job-runner.toml (hunk 5 above)
Comment (martinpitt) [reply job-runner.toml:56]: Tracked in #6030 now. However, is that actually hard to fix? It's already commented out after all, and you said that this is the internal defaults -- so doesn't that mean that there is no default, and it somehow works without anyway?

If it's hard, then let's do it as a follow-up, but I'd like to at least understand why.

Comment (allisonkarlitskaya) [reply job-runner.toml:56]: It's absolutely not difficult to do, at least for the GitHub token.  The S3 stuff is slightly more complicated since we'd have to bring back the dir-scanning/suffix-matching logic.

For me it's more a question of the reasonableness of the default.  Do you picture a situation where we should use job-runner locally on real PRs and have statuses posted from our own GitHub accounts?  That's effectively what you're requesting here.

I think the "broken by default" default config is a good thing...

Comment (martinpitt) [reply job-runner.toml:56]: I was missing that you *intentionally* wanted to "break" the default. So this stops being a blocker for this PR for me, and we can continue to discuss it in #6030.

My worry is more that the intended mode of operation requires duplicating my github token, it has to be both in ~/.config and in that toml. That makes rotating tokens harder, and I like "single source of truth".

Comment (allisonkarlitskaya) [reply job-runner.toml:56]: For the reason above, *if* we wanted to solve the broken-by-default thing, I'd be fairly insistent about solving it in the `forge.github.post = false` direction.

File: lib/aio/s3.py [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+        for attempt in range(11):
+            try:
+                return await self.request_once(request, checksum)
+            except aiohttp.ClientResponseError as exc:
+                if exc.status < 500:
+                    raise  # We actually care about 4xx errors, but blindly retry 5xx.
+            except aiohttp.ClientError:
+                pass  # That's DNS, connection, etc. errors.  Always retry those.
+
+            # 0.125 → 0.25 → 0.5 → 1 → 4 → 16 → 32s back-off
```
Comment (martinpitt) [inline lib/aio/s3.py:75]: This feels too aggressive -- in many cases we got "50x slow down", and retrying after milliseconds is kind of the opposite. Maybe we need to special-case "slow down", but even for DNS or routing errors (overload? congestion?) they have a better chance with less aggressive retrying.

File: lib/aio/util.py [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+        self.append(item)
+        self._nonempty.set()
+
+    def eof(self) -> None:
+        self._nonempty.set()
+        self._eof = True
+
+
+# Simple LRU cache: when full, evict the least-recently `add()`-ed item.
+class LRUCache(dict[K, V]):
```
Comment (martinpitt) [inline lib/aio/util.py:63]: How is that different from using https://docs.python.org/3/library/functools.html#functools.lru_cache directly in github.py?

File: .github/workflows/test.yml (hunk 23 above)
Comment (allisonkarlitskaya) [reply .github/workflows/test.yml:22]: I kinda want to keep this... after all, it's "free".  Also no strong opinion, though.

File: lib/aio/job.py [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+            f'--cidfile={cidfile}',
+            *(f'--env={k}={v}' for k, v in job.env.items()),
+            '--env=TEST_ATTACHMENTS=/tmp/attachments',
+            f'--env=COCKPIT_CI_LOG_URL={log.url}',
+            *itertools.chain.from_iterable(args for name, args in ctx.secrets_args.items() if name in job.secrets),
+
+            job.container or ctx.default_image,
+
+            'python3', '-c', Path(f'{BOTS_DIR}/checkout-and-run').read_text(),  # lulz
+            f'--revision={subject.sha}'
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py:96]: OK.

File: lib/aio/job.py [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+            try:
+                # Log data until we hit EOF
+                assert container.stdout is not None
+                async for block in read_utf8(container.stdout):
+                    log.write(block)
+
+                # Upload all attachments
+                # TODO: live updates
+                # TODO: invent async tarfile for StreamReader
+                with contextlib.suppress(FileNotFoundError):
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py:116]: OK.

File: .github/workflows/test.yml (hunk 23 above)
Comment (martinpitt) [reply .github/workflows/test.yml:22]: Yeah, doesn't hurt.

File: lib/aio/job.py [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+                    cid = cidfile.read_text().strip()
+                    await run([*ctx.container_cmd, 'cp', '--', f'{cid}:/tmp/attachments/.', f'{attachments}'])
+                    for file in attachments.rglob('*'):
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py:119]: I don't expect to attach dotfiles and we can't upload directories to S3.  Unless you mean the files inside of the directories and then: note the `r` in `rglob()`.

File: lib/aio/job.py [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+                        log.index.write(str(file.relative_to(attachments)), file.read_bytes())
+
+                if returncode := await container.wait():
+                    raise Failure(f'Container exited with code {returncode}')
+
+            finally:
+                await run([*ctx.container_cmd, 'rm', '--force', '--time=0', f'--cidfile={tmpdir}/cidfile'],
+                          stdout=asyncio.subprocess.DEVNULL, stderr=asyncio.subprocess.DEVNULL)
+
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py:128]: I think I added this redirection before removing `--rm` on the container.  At that time, it was meant as a mop-up-in-case-of-crash rather than the main way of killing the container, and would give the error message about the container missing in the usual case.  That's no longer true, so I'll fix it.  Thanks for the suggestion.

File: lib/aio/s3.py (hunk 24 above)
Comment (allisonkarlitskaya) [reply lib/aio/s3.py:75]: I feel like in most cases the error will be resolved by trying again immediately, but if it's not, then we'll slowly back off.

Note that the S3 retry stuff is much more aggressive than GitHub.  We have a much higher limit on S3 (we're paying them, after all)...

There's also no real harm in backing this off either.  Can you make a suggestion about the progression (or algorithm) you'd like to see?

File: lib/aio/util.py (hunk 25 above)
Comment (allisonkarlitskaya) [reply lib/aio/util.py:63]: @functools.lru_cache applied to the `get()` method would have two problems:
 - @functools.lru_cache() on instance methods leaks (since the self parameter gets included as one of the "inputs").  Probably not relevant for us, since we have only one of them; but
 - this means that repeated `get()` calls on the same resource would always return the same value without bothering to check if it changed.  We actually *do* want to do the http request — we're polling after all — just with the etag/last-modified headers set.

File: lib/aio/job.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/job.py:119]: Ah, TIL `rglob`.

File: lib/aio/job.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/job.py:128]: In, fact, why didn't you just use `run --rm`? Isn't that easier and more robust?

File: lib/aio/s3.py (hunk 24 above)
Comment (martinpitt) [reply lib/aio/s3.py:75]: Gut feeling: Lower retry to 5, and `sleep(2 ** (2 * attempt)`. That gives 1 → 4 → 16 → 64 → 256. It's not a hard science of course.

Note that this is less about paying, and more about the practictalities of networking :grin:

File: lib/aio/util.py (hunk 25 above)
Comment (martinpitt) [reply lib/aio/util.py:63]: Sure, I wasn't proposing to apply this to the get() method, but to some `lru_cache(self.cache.__get__)` wrapping. But I did a quick google search and this doesn't actually seem to be that simple. Too bad, that feels like it'd be a nice stdlib addition.

File: job-runner.toml [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+[container.secrets]
+# see podman-secret(1)
+# github-token = ['--secret=github-token']
+
+[logs]
+driver='local'   # 's3' or 'local'
+
+[logs.s3]
+# hint: podman run --rm --net=host quay.io/minio/minio server /var
+url = 'http://127.0.0.1:9000/tmp/'
```
Comment (martinpitt) [inline job-runner.toml:37]: This is a bit awkward -- production has to duplicate `LOG_STORE` from lib/stores.py. Commenting this out and defaulting to `LOG_STORE` would be nicer.

Comment (martinpitt) [reply job-runner.toml:37]: But nevermind, that'd go against the "become independent from bots" idea.

File: lib/aio/job.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py:128]: The container needs to survive the test run exiting in order for us to retrieve the attachments.

This is tricky, and something we talked about before.  Even if we move to a model where we support live attachments again, we face the possibility of an attachment being written at the end of a test run and the container immediately exiting.

That implies that we need to accept one of the following:
 - we need a streaming protocol that's part of the data that the main process of the container delivers before it exits, either via stdout/err or some sort of fd hacks; or
 - we need attachments stored in a volume that survives the container `rm`; or
 - we can't use `--rm`.

File: lib/aio/s3.py (hunk 24 above)
Comment (allisonkarlitskaya) [reply lib/aio/s3.py:75]: OK.

File: lib/aio/job.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/job.py:128]: Of all these, creating a `temp_dir/attachments` on the host and mounting that as a volume feels best to me. That part can be a follow-up.

[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py]: I actually like the idea of having attachments as something done more explicitly via testlib API, with strong assurances that everything is dealt with by the time the main process exits (ie: no longer based on having a directory at all, and not involving the possibility of having stuff hanging around after the container exits).  This starts sounding like tar-via-passed-in-fd or so, but maybe we could find another way...

[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply lib/aio/job.py]: Straw man:
 - `test/common/run-tests` checks if `TEST_ATTACHMENT_FD` is set, and if so, creates a temporary directory and sets the `TEST_ATTACHMENTS` envvar pointing to it
 - on exit, it calls `tar` on the contents of that directory directing output to the fd given in the envvar
 - from job-runner we use tarfile directly into s3streamer and nothing ever touches the disk outside of the container
 - we can still use the `TEST_ATTACHMENTS` directory approach in places where fd passing isn't supported

Definitely a follow-up, though.  I realized quite early in working on this that the millions of ways to handle attachments would be a time-sucking pit and quickly decided to slap the `# TODO` comment on it and move on!!

[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/job.py]: Agreed to postpone this. That'll be a longer discussion, and I'll accept the non-live attachments regression for now.

[the remark this answers is not in the document]
Comment (martinpitt) [reply lib/aio/job.py]: Ah, you already fixed that in the latest force-push.

**Review by martinpitt — APPROVED** [review]
Let's do this! :star: :rocket: :checkered_flag:
